core/src/SYCL/Kokkos_SYCL_Team.hpp

```diff
@@ -0,0 +1,627 @@
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_SYCL_TEAM_HPP
+#define KOKKOS_SYCL_TEAM_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#ifdef KOKKOS_ENABLE_SYCL
+
+#include <utility>
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename Type>
+struct SYCLJoinFunctor {
+  using value_type = Type;
+
+  KOKKOS_INLINE_FUNCTION
+  static void join(volatile value_type& update,
+                   volatile const value_type& input) {
+    update += input;
+  }
+};
+
+/**\brief  Team member_type passed to TeamPolicy or TeamTask closures.
+ */
+class SYCLTeamMember {
+ public:
+  using execution_space      = Kokkos::Experimental::SYCL;
+  using scratch_memory_space = execution_space::scratch_memory_space;
+
+ private:
+  mutable void* m_team_reduce;
+  scratch_memory_space m_team_shared;
+  int m_team_reduce_size;
+  int m_league_rank;
+  int m_league_size;
+  sycl::group<1> m_group;
+
+ public:
+  KOKKOS_INLINE_FUNCTION
+  const execution_space::scratch_memory_space& team_shmem() const {
+    return m_team_shared.set_team_thread_mode(0, 1, 0);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  const execution_space::scratch_memory_space& team_scratch(
+      const int& level) const {
+    return m_team_shared.set_team_thread_mode(level, 1, 0);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  const execution_space::scratch_memory_space& thread_scratch(
+      const int& level) const {
+    return m_team_shared.set_team_thread_mode(level, team_size(), team_rank());
+  }
+
+  KOKKOS_INLINE_FUNCTION int league_rank() const { return m_league_rank; }
+  KOKKOS_INLINE_FUNCTION int league_size() const { return m_league_size; }
+  KOKKOS_INLINE_FUNCTION int team_rank() const {
+    // FIXME_SYCL should be m_group.get_local_id() in SYCL2020
+    Kokkos::abort("Not implemented!");
```

Comment:
could we compute a rank via something like this in the constructor of the team handle:
```
parallel_for(TeamThreadRange(), 0, 10000) {
if(m_team_rank==0) {
  m_team_rank = atomic_fetch_add(&counter_array[league_rank()],1);
});
```
where counter_array is some array of league_size length?

Reply:
if we do this we can also fix the single statement. If you say you want to do it in another PR ok.

Reply:
Something like
```C++
 auto count_team_rank_ptr = static_cast<int*>(
        sycl::malloc(sizeof(int)*m_league_size, q, sycl::usm::alloc::device));

    q.submit([&](sycl::handler& cgh) {
      const auto shmem_begin  = m_shmem_begin;
      const auto shmem_size   = m_shmem_size;
      const auto scratch_size = m_scratch_size[1];
      const auto team_size    = m_team_size;
      auto team_lambda        = [=](sycl::group<1> g) {
        int my_team_rank = 0;
        int* my_team_rank_ptr = &my_team_rank;
        auto local_count_team_rank_ptr = &count_team_rank_ptr[g.get_id(0)];
        *local_count_team_rank_ptr = 0;
        KOKKOS_IMPL_DO_NOT_USE_PRINTF("team_rank out: %d %p\n", *my_team_rank_ptr, my_team_rank_ptr);
        g.parallel_for_work_item(sycl::range<1>(team_size),
          [=](sycl::h_item<1>) {
            KOKKOS_IMPL_DO_NOT_USE_PRINTF("team_rank in1: %d %p\n", *my_team_rank_ptr, my_team_rank_ptr);
            *my_team_rank_ptr = atomic_fetch_add(local_count_team_rank_ptr,1);
            KOKKOS_IMPL_DO_NOT_USE_PRINTF("team_rank in1a: %d\n", *local_count_team_rank_ptr);
            KOKKOS_IMPL_DO_NOT_USE_PRINTF("team_rank in2: %d %p\n", *my_team_rank_ptr, my_team_rank_ptr);
          });

        // FIXME_SYCL Add scratch memory
        member_type team_member(nullptr, shmem_begin, shmem_size, nullptr,
                                      scratch_size, g);
        team_member.impl_set_team_rank(my_team_rank);
[...]
```
results in
```
team_rank out: 0 0x7ffd60b0d600
team_rank in1: 0 0x7ffd60b0d600
team_rank in1a: 1
team_rank in2: 0 0x7ffd60b0d600
team_rank in1: 0 0x7ffd60b0d600
team_rank in1a: 2
team_rank in2: 1 0x7ffd60b0d600
tR: 1, lR: 0
team_rank out: 0 0x7f518ff9be00
team_rank in1: 0 0x7f518ff9be00
team_rank in1a: 1
team_rank in2: 0 0x7f518ff9be00
team_rank in1: 0 0x7f518ff9be00
team_rank in1a: 2
team_rank in2: 1 0x7f518ff9be00
tR: 1, lR: 2
team_rank out: 0 0x7f518fb9ae00
team_rank in1: 0 0x7f518fb9ae00
team_rank in1a: 1
team_rank in2: 0 0x7f518fb9ae00
team_rank in1: 0 0x7f518fb9ae00
team_rank in1a: 2
team_rank in2: 1 0x7f518fb9ae00
tR: 1, lR: 1
```
for https://github.com/kokkos/kokkos/blob/develop/core/unit_test/incremental/Test10_HierarchicalBasics.hpp. In particular, the outer loop is executed once per team and not once per thread.